### ISSUE TITLE
Potential regression in deterministic outputs

### ISSUE THREAD

**Issue by pcuenca** [huggingface/diffusers#902]
### Describe the bug

I've started noticing different outputs ~~in the latest version of diffusers~~ starting from diffusers `0.4.0` when compared against `0.3.0`. This is my test code (extracted from a notebook):

```Python
import diffusers
from diffusers import StableDiffusionPipeline, DDIMScheduler
import torch
from IPython.display import display

def run_tests(pipe):
    torch.manual_seed(1000)
    display(pipe("A photo of Barack Obama smiling with a big grin").images[0])
    torch.manual_seed(1000)
    display(pipe("Labrador in the style of Vermeer").images[0])

pipe = StableDiffusionPipeline.from_pretrained("CompVis/stable-diffusion-v1-4")
pipe = pipe.to("cuda")
run_tests(pipe)
```

The first prompt produces identical results. The second one, however, results in different outputs:

`0.3.0`
![labrador_0 3](https://user-images.githubusercontent.com/1177582/196655442-3117bed1-abc9-41d6-a813-6eb7044be2b2.png)

`main@a3efa433eac5feba842350c38a1db29244963fb5`
![labrador_0 6](https://user-images.githubusercontent.com/1177582/196655457-af371112-c81e-42f2-a9ba-56042973988b.png)

Using DDIM, both prompts generate different images.

```Python
scheduler = DDIMScheduler(beta_start=0.00085, beta_end=0.012, beta_schedule="scaled_linear", num_train_timesteps=1000)
pipe = StableDiffusionPipeline.from_pretrained("CompVis/stable-diffusion-v1-4", scheduler=scheduler)
pipe = pipe.to("cuda")
run_tests(pipe)
```

`DDIM 0.3.0`
![obama_ddim_0 3](https://user-images.githubusercontent.com/1177582/196656145-46310f7b-15d5-44c6-bb2a-3c085ded948a.png)

`DDIM main`
![obama_ddim_0 6](https://user-images.githubusercontent.com/1177582/196656180-670ac3fd-960a-4aa8-bc01-a9f566d1bb44.png)

`DDIM 0.3.0`
![labrador_ddim_0 3](https://user-images.githubusercontent.com/1177582/196656238-0fa6d704-a96a-408f-8f1c-b33cb100a837.png)

`DDIM main`
![labrador_ddim_0 6](https://user-images.githubusercontent.com/1177582/196656259-f42ed55c-e667-4a4b-9733-fd46f78f64b4.png)

In addition, there's this post from a forum user with very different results in the img2img pipeline: https://discuss.huggingface.co/t/notable-differences-between-other-implementations-of-stable-diffusion-particularly-in-the-img2img-pipeline/24635/5. They opened another issue recently #901. Cross-referencing, may or may not be related to this issue.

### Reproduction

As explained above.

### System Info

diffusers: `main` @ `a3efa433eac5feba842350c38a1db29244963fb5` vs `v0.3.0`

**Comment by pcuenca** [comment]
Update regarding the `img2img` pipeline: there are small differences using DDIM, I didn't notice any with PNDM or k-LMS. All tests done in full precision for now.

`DDIM v0.3.0`
![i2i_ddim_0 3](https://user-images.githubusercontent.com/1177582/196667678-b53864c8-9019-4503-a9ba-852be46875c0.png)

`DDIM main`
![i2i_ddim_0 6](https://user-images.githubusercontent.com/1177582/196667690-d08c2c68-1ede-4a9e-a136-649d7a465889.png)

**Comment by patrickvonplaten** [comment]
Thanks a lot @pcuenca ! 

Do you think we could try to narrow down even more when the differences started happening? E.g. I guess it's between 0.3.0 and 0.4.0 , but then it'd also be nice to find out exactly which PR is the origin :-) 

Otherwise happy to look into it myself in ~2 days!

**Comment by pcuenca** [comment]
Yes, I'll investigate a bit :)

**Comment by cccntu** [comment]
I tried it on `0.3.0` and some `0.5.0` fork I have. And in both version the dog looks like the second one.

**Comment by pcuenca** [comment]
> I tried it on `0.3.0` and some `0.5.0` fork I have. And in both version the dog looks like the second one. Can you reproduce it with docker?

That's interesting. I haven't tried docker, but all the tests were done in the same system just updating the version of `diffusers`.

**Comment by ydshieh** [comment]
@pcuenca Could you provide the system info? GPU type, python/torch/torchvision versions, CUDA/CUDNN versions?

**Comment by ydshieh** [comment]
I can't reproduce with T4 GPU on GCP. Both commits give the second dog image.

**Comment by ydshieh** [comment]
For dog prompt with DDIM: running from commit `3f55d13` to `9bca402`, there are 4 commits where image difference occurs

074-a9fdb3d  => 075-6bd005e:  0.09812037150065105
099-3b747de =>100-bd8df2d:   0.0812060038248698
116-a7058f4  => 117-9ebaea5:   0.22810236612955728
132-1070e1a  =>133-f1484b8:    0.17296854654947916

The last number is the ratio of equal values over the total number of values in the numpy arrays
I upload the images to [this repo](https://huggingface.co/datasets/ydshieh/temp-dog-diff/tree/main).

For the default scheduler, also these 4 commits give some difference in the numpy arrays, but the ratio > 0.998, so visually we can't see any difference.

**Comment by ydshieh** [comment]
![dog](https://user-images.githubusercontent.com/2521628/197385414-eb52951d-4e17-47dc-bc1f-611580d91b55.png)

**Comment by pcuenca** [comment]
Hi @ydshieh, thanks a lot for looking into this! I've done a few more experiments and these are my results:

<details>
<summary>My workstation</summary>

```
- GPU: 3090, CUDA 11.4
- Platform: Linux-5.4.0-91-generic-x86_64-with-glibc2.31
- Python version: 3.9.12
- PyTorch version (GPU?): 1.12.1+cu113 (True)
- Huggingface_hub version: 0.10.0
- Transformers version: 4.22.2
```

`0.3.0 fp32`
![0 3-fp32-3090](https://user-images.githubusercontent.com/1177582/197487849-dc34f4ec-ae77-4b49-a009-191dce7fcf44.png)

`0.6.0 fp16` (Note different ear)
![0 6-fp16-3090](https://user-images.githubusercontent.com/1177582/197487857-616d3cad-a0d6-4499-a054-749a8e5d819b.png)

`0.6.0 fp32` (Blue hat)
![0 6-fp32-3090](https://user-images.githubusercontent.com/1177582/197487865-387635e7-fe69-4f52-97c1-d1cd3f8aa6bf.png)

</details>

<details>
<summary>Colab T4</summary>


```
- GPU: Tesla T4, CUDA 11.2
- Platform: Linux-5.10.133+-x86_64-with-Ubuntu-18.04-bionic
- Python version: 3.7.15
- PyTorch version (GPU?): 1.12.1+cu113 (True)
- Huggingface_hub version: 0.10.1
- Transformers version: 4.23.1
```

`0.3.0 fp32`
![0 3-fp32-colab-t4](https://user-images.githubusercontent.com/1177582/197488560-64337089-e406-40b0-9d7f-aca39df65a48.png)

`0.6.0 fp16`
![0 6-fp16-colab-t4](https://user-images.githubusercontent.com/1177582/197488568-8fffc9dd-4537-4fb3-8ddd-7351226e47d4.png)

`0.6.0 fp32`
![0 6-fp32-colab-t4](https://user-images.githubusercontent.com/1177582/197488572-483f6c77-7aa0-4511-bb17-20df5bc8d00a.png)

</details>

<details>
<summary>Colab A100</summary>

```
- GPU: A100 (40 GB) CUDA 11.2
- Platform: Linux-5.10.133+-x86_64-with-Ubuntu-18.04-bionic
- Python version: 3.7.15
- PyTorch version (GPU?): 1.12.1+cu113 (True)
- Huggingface_hub version: 0.10.1
- Transformers version: 4.23.1
```

`0.3.0 fp32`
![0 3-fp32-colab-a100](https://user-images.githubusercontent.com/1177582/197488931-d9b9d6e9-1e84-49ea-ae17-fd65d3ccabb9.png)

`0.6.0 fp16`
![0 6-fp16-colab-a100](https://user-images.githubusercontent.com/1177582/197488945-609f06a6-f228-4483-86cd-569455f9c65c.png)

`0.6.0 fp32`
![0 6-fp32-colab-a100](https://user-images.githubusercontent.com/1177582/197488950-70f5faaa-b8d0-40bc-a0c7-3f5cb59bc5f8.png)

</details>

As you can see, results differ across versions. Even if we ignore the `fp16` tests, results in the A100 Colab are different to those in the T4 environment. The A100 fp32 versions are consistent between diffusers 0.3.0 and 0.6.0 (no blue hat), and T4 fp32 are also consistent but different (blue hat). In my computer, 0.3.0 has no blue hat but 0.6.0 has.

Any idea about what might be going on?

**Comment by ydshieh** [comment]
@pcuenca No useful insight from my side (at least so far).

In general, I am not very sure if it really makes sense to ensure reproducibility across different GPUs. 

Regarding the difference across versions, I think we can take a look at the 4 commits I mentioned above. Notice that `diffusers` have (a lot of?) optimizations, which are probably the causes of the difference.

IMO, we should try to ensure the reproducibility on a fix environment but across `diffusers` versions. And if we do have some commit that will introduce the difference, it should be clearly recorded, so we can find that information quickly whenever necessary.

**Comment by patrickvonplaten** [comment]
Small update here:
- 1.) We now know that we cannot guarantee reproducibility (only loosely "close" reproducibility) because of https://github.com/pytorch/pytorch/issues/87992 => therefore we can never really guarantee that the exact same images are generated across devices
- 2.) I checked and I cannot reproduce difference of this code:
```python
import diffusers
from diffusers import StableDiffusionPipeline, DDIMScheduler
import torch
from IPython.display import display

def run_tests(pipe):
    torch.manual_seed(1000)
    display(pipe("A photo of Barack Obama smiling with a big grin").images[0])
    torch.manual_seed(1000)
    display(pipe("Labrador in the style of Vermeer").images[0])

pipe = StableDiffusionPipeline.from_pretrained("CompVis/stable-diffusion-v1-4")
pipe = pipe.to("cuda")
run_tests(pipe)
```
between 0.3.0 and 0.7.0dev using a V100
- 3.) The aggressive unittests: https://github.com/huggingface/diffusers/blob/82d56cf192f3a3c52e0708b6c8db4a6d959244dd/tests/models/test_models_unet_2d.py#L414 all pass for 0.3.0
This is good as it means our unet is not responsible for the potential regression above


Overall this issue to me now seems much less severe than originally and a bit part of it is probably simply to "uncontrollable" randomness

Next:
- Add aggressive scheduler tests and check differences between 0.3.0 and 0.7.0dev
- Add aggressive minimal step pipeline tests and check differences between 0.3.0 and 0.7.0dev

**Comment by ydshieh** [comment]
Just curious: 

> I checked and I cannot reproduce difference of this code:

What kind of difference you are checking/looking here, @patrickvonplaten ?

**Comment by patrickvonplaten** [comment]
Once the pipeline tests are fully updated we should also make a doc explaining the problem with reproducibility in general with diffusion models.
